pkg/3scale/amp/operator/base_apimanager_logic_reconciler.go

```diff
@@ -116,6 +133,269 @@ func (r *BaseAPIManagerLogicReconciler) ReconcileGrafanaDashboard(desired *grafa
+
+    // system-sidekiq pod needs to be up & running
+    err := r.waitForSystemSidekiq(apimanager)
+    if err != nil {
+        return "", fmt.Errorf("failed to wait for system-sidekiq: %w", err)
+    }
+
+	listOps := []client.ListOption{
+		client.InNamespace(namespace),
+		client.MatchingLabels(map[string]string{"deploymentConfig": "system-sidekiq"}),
+	}
+
+	err = r.Client().List(context.TODO(), podList, listOps...)
+	if err != nil {
+		return "", fmt.Errorf("failed to list pods: %w", err)
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Status.Phase == "Running" {
+			podName = pod.ObjectMeta.Name
+			break
+		}
+	}
+	
+	if podName == "" {
+		return "",fmt.Errorf("no matching pod found")
+	}
+
+	return podName, nil
+}
+
+func (r *BaseAPIManagerLogicReconciler) getAccessToken() (string, error) {
+
+    namespace := r.apiManager.GetNamespace()
+
+	secretName := types.NamespacedName{
+        Namespace: namespace,
+        Name:      "system-seed",
+    }
+
+    secret := &v1.Secret{}
+    err := r.Client().Get(context.TODO(), secretName, secret)
+    if err != nil {
+        return "", fmt.Errorf("failed to get secret: %w", err)
+    }
+
+    // Retrieve the access token from the secret data
+    accessToken, ok := secret.Data["MASTER_ACCESS_TOKEN"]
+    if !ok {
+        return "", fmt.Errorf("access token not found in secret")
+    }
+
+    return string(accessToken), nil
+}
+
+func (r *BaseAPIManagerLogicReconciler) getMasterRoute() (string, error) {
+
+	masterPrefix := "master"
+
+	opts := []client.ListOption{
+        client.InNamespace(r.apiManager.GetNamespace()),
+    }
+	foundRoute := ""
+	routes := routev1.RouteList{}
+    err := r.Client().List(context.TODO(), &routes, opts ...)
+    if err != nil {
+        return "", err
+    }
+
+	for _, route := range routes.Items {
+		if strings.HasPrefix(route.Spec.Host, masterPrefix) {
+			foundRoute = "https://" + route.Spec.Host
+			return foundRoute, nil
+		}
+	}
+
+	return "", fmt.Errorf("route not found")
+}
+
+func (r *BaseAPIManagerLogicReconciler) baseRoutesExist() (bool, error) {
+	expectedRoutes := 2
+    serviceNames := []string{"system-master", "backend-listener"}
+    opts := []client.ListOption{
+        client.InNamespace(r.apiManager.GetNamespace()),
+    }
+
+    routes := routev1.RouteList{}
+    err := r.Client().List(context.TODO(), &routes, opts ...)
+    if err != nil {
+        return false, err
+    }
+
+    serviceCount := 0
+    for _, service := range serviceNames {
+        found := false
+        for _, route := range routes.Items {
+            if route.Spec.To.Name == service {
+                found = true
+                break
+            }
+        }
+        if found {
+            serviceCount++
+        }
+    }
+
+    if serviceCount >= expectedRoutes {
+        return true, nil
+    }
+
+    return false, fmt.Errorf("base routes not found")
+}
+
+func (r *BaseAPIManagerLogicReconciler) getAccountUrls() ([]string, []string, error){
```

Comment:
There is a library where we have this implemented: https://github.com/3scale/3scale-porta-go-client